jetty-core/jetty-websocket/jetty-websocket-jetty-server/src/main/java/org/eclipse/jetty/websocket/server/internal/CompletedUpgradeRequest.java

```diff
@@ -0,0 +1,200 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.server.internal;
+
+import java.net.HttpCookie;
+import java.net.URI;
+import java.security.Principal;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.jetty.http.HttpField;
+import org.eclipse.jetty.http.HttpFields;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpScheme;
+import org.eclipse.jetty.http.HttpURI;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.util.Fields;
+import org.eclipse.jetty.websocket.api.ExtensionConfig;
+import org.eclipse.jetty.websocket.api.UpgradeRequest;
+import org.eclipse.jetty.websocket.common.JettyExtensionConfig;
+import org.eclipse.jetty.websocket.core.server.ServerUpgradeRequest;
+
+public class CompletedUpgradeRequest implements UpgradeRequest
+{
+    private final HttpFields _httpFields;
+    private final List<HttpCookie> _cookies;
+    private final List<ExtensionConfig> _extensions;
+    private final HttpURI _httpURI;
+    private final String _httpVersion;
+    private final String _method;
+    private final List<String> _subProtocols;
+    private final Map<String, List<String>> _queryParams;
+    private final boolean _secure;
+    private final String _protocolVersion;
+    private final Principal _userPrincipal;
+
+    public CompletedUpgradeRequest(ServerUpgradeRequest request)
+    {
+        _httpFields = request.getHeaders().asImmutable();
+        _httpURI = request.getHttpURI();
+        _method = request.getMethod();
+        _httpVersion = request.getConnectionMetaData().getHttpVersion().asString();
+        _subProtocols = request.getSubProtocols();
+        _secure = request.isSecure();
+        _protocolVersion = request.getProtocolVersion();
+
+        // Convert Jetty Cookies to java.net.HttpCookie.
+        List<HttpCookie> cookies = new ArrayList<>();
+        for (org.eclipse.jetty.http.HttpCookie cookie : Request.getCookies(request))
+        {
+            cookies.add(org.eclipse.jetty.http.HttpCookie.asJavaNetHttpCookie(cookie));
+        }
+        _cookies = Collections.unmodifiableList(cookies);
```

Comment:
Could this be done lazily?  I think Jetty Cookies are immutable, so you can keep a reference to them